### ISSUE TITLE
Data compression DiskDataset

### ISSUE THREAD

**Issue by ashtawy** [deepchem/deepchem#3088]
## 🚀 Feature

DiskDataset is very helpful when dealing with large datasets that do not fit in memory. However, the stored data, specifically high-dimensional X NumPy arrays of each shard, take up a lot of disk space when screening very large libraries. It would be great to support the compressed `.npz` format using numpy.savez_compressed and make it the default format instead of the vanilla numpy.save

Links:
https://numpy.org/doc/stable/reference/generated/numpy.savez_compressed.html
https://github.com/deepchem/deepchem/blob/2.6.1/deepchem/data/datasets.py#L1331
https://github.com/deepchem/deepchem/blob/2.6.1/deepchem/utils/data_utils.py#L469

**Comment by peastman** [comment]
Assuming the compression doesn't hurt loading speed, this seems like a good idea.  If it does make loading slower, it should probably be an option but not the default.

**Comment by ashtawy** [comment]
@peastman agreed regarding saving/loading speed and exposing compression as an optional feature in DiskDataset. However, I haven't noticed a significant difference between `npy` and `npz` files when saving/loading several datasets in the past. Time spent saving the featurized dataset is usually a fraction of the featurization time, so it is effectively negligible. One can profile the loading time to decide whether the npz option should be the default or not.

**Comment by rbharath** [comment]
CC @arunppsg who has been also thinking of ways to lower storage requirements for large `DiskDataset` objects.

@ashtawy Nice suggestion! Do you have a sense of how big the space difference is for .npz files instead of .npy files?

**Comment by ashtawy** [comment]
Generally, the saving is a function of the data (entropy) since compression algorithms take advantage of data repetition to reduce the number of bits required to save the data. 

For drug-discovery applications, the saving can be more than 30X cheaper when dealing with binary data (molecular fingerprints are binary most of the time and they can be binarized/bucketed if they are continues to make model training more stable). This can literally save us 30X in $ when storing our data on AWS S3 or EFS.

<img width="630" alt="image" src="https://user-images.githubusercontent.com/3431058/197365293-a743c950-651e-4c75-b885-e295b5ea536c.png">

**Comment by gehlhaarPfizer** [comment]
This is important enough to us that we will implement this ourselves if it is not added to DeepChem.

**Comment by rbharath** [comment]
@gehlhaarPfizer If you are able to, we would very much appreciate a contribution to DeepChem.

CC @ARY2260. The large storage space on Disk issue has come up for us too in different contexts so we should reprioritize this

**Comment by gehlhaarPfizer** [comment]
@rbharath I will likely be unable to officially join the project but I am happy to email you or upload code, deltas. In my own work I found numpy.savez_compressed to be annoying, as it is based on (for no reason) saving a list of arrays that need to be parsed; and that the easiest solution by far is to use gzip.GzipFile and pass that into numpy.save. In any case the code delta is just a few lines. I can work on that in the next week or two and get it to you.

**Comment by gehlhaarPfizer** [comment]
I have implemented this, along with changes that resolve #4059 and #4060. Please see attached.
[datasets.zip](https://github.com/user-attachments/files/16269290/datasets.zip)
